File: tests/cases/objects.py
"""Oracle-diff cases for poketcg/src/home/objects.asm."""

POISON = {"a": 0xAA, "f": 0xF0, "b": 0xBB, "c": 0xCC, "d": 0xDD, "e": 0xEE, "hl": 0x1234}

wOAM = 0xCA00
OAM_SIZE = 160
wOAMOffset = 0xCAB5

CONTRACT = {
    "SetOneObjectAttributes": ("b", "c", "d", "e", "hl"),
    "ZeroObjectPositions": ("b", "d", "e"),
    "SetManyObjectsAttributes": ("f", "d", "e", "hl"),
}

OBJ = bytes((0x11, 0x22, 0x33, 0x44)) * 40
SPRITE = bytes((0x11, 0x22, 0x33, 0x44))
SRC40 = bytes((0,)) + SPRITE * 40  # count 0 means 256; only 40 fit before OAM fills
SRC1 = bytes((1,)) + SPRITE
SRC = 0xC100

CASES = {
    "SetOneObjectAttributes": [
        {"e": 0x10, "d": 0x20, "c": 0x30, "b": 0x40, "wram": {wOAMOffset: b"\x00"}, "read": {wOAM: 4, wOAMOffset: 1}},
        dict(POISON, e=0x50, d=0x60, c=0x70, b=0x80, wram={wOAMOffset: b"\x08"}, read={wOAM + 8: 4, wOAMOffset: 1}),
        {"e": 0xA5, "d": 0x5A, "c": 0x01, "b": 0x02, "wram": {wOAMOffset: b"\x9C"}, "read": {wOAM + 0x9C: 4, wOAMOffset: 1}},
        {"wram": {wOAMOffset: bytes((OAM_SIZE,))}, "read": {wOAM: 4, wOAMOffset: 1}},
    ],
    "ZeroObjectPositions": [
        {"read": {wOAM: OAM_SIZE, wOAMOffset: 1}},
        dict(POISON, wram={wOAM: OBJ, wOAMOffset: b"\x28"}, read={wOAM: OAM_SIZE, wOAMOffset: 1}),
    ],
    "SetManyObjectsAttributes": [
        # All-zero registers: hl=0 reads real ROM bytes at $0000, off defaults 0.
        # OAM fills at 40 sprites regardless of how large that count byte is.
        {"read": {wOAM: OAM_SIZE, wOAMOffset: 1}},
        # n==0 means 256: off=0 still bails exactly at 40 sprites (160 bytes).
        {"hl": SRC, "d": 0x20, "e": 0x10, "wram": {SRC: SRC40, wOAMOffset: b"\x00"},
         "read": {wOAM: OAM_SIZE, wOAMOffset: 1}},
        {"hl": SRC, "wram": {SRC: SRC1, wOAMOffset: b"\x00"},
         "read": {wOAM: 4, wOAMOffset: 1}},
        dict(POISON, hl=SRC, d=0x05, e=0x03,
             wram={SRC: bytes((2,)) + SPRITE * 2, wOAMOffset: b"\x08"},
             read={wOAM + 8: 8, wOAMOffset: 1}),
        # Exact boundary: one sprite lands right at OAM_SIZE (carry, Z set).
        {"hl": SRC, "wram": {SRC: SRC1, wOAMOffset: bytes((OAM_SIZE - 4,))},
         "read": {wOAM + OAM_SIZE - 4: 4, wOAMOffset: 1}},
        # Off-by-one past OAM_SIZE (carry, Z clear).
        {"hl": SRC, "wram": {SRC: SRC1, wOAMOffset: bytes((OAM_SIZE - 3,))},
         "read": {wOAM + OAM_SIZE - 3: 4, wOAMOffset: 1}},
        # Initial reject, offset exactly OAM_SIZE.
        {"hl": SRC, "wram": {wOAMOffset: bytes((OAM_SIZE,))}, "read": {wOAMOffset: 1}},
        # Initial reject, offset past OAM_SIZE.
        {"hl": SRC, "wram": {wOAMOffset: b"\xC8"}, "read": {wOAMOffset: 1}},
    ],
}
from tests.cases._schema_migration import legacy_to_schema
SCHEMA2_CASES = legacy_to_schema(CASES, CONTRACT)

File: src/home/objects.c
#include "home/objects.h"

#include "generated/wram.h"
#include "mem.h"

#define OAM_COUNT 40u
#define OAM_SIZE  (OAM_COUNT * 4u)

void SetOneObjectAttributes(uint8_t e, uint8_t d, uint8_t c, uint8_t b)
{
	uint8_t off = wOAMOffset;
	if (off >= OAM_SIZE)
		return;
	uint16_t p = (uint16_t)(wOAM_ADDR + off);
	gb_write8(p++, e);
	gb_write8(p++, d);
	gb_write8(p++, c);
	gb_write8(p, b);
	wOAMOffset = (uint8_t)(off + 4u);
}

/* objects.asm:72-85. Zeroes only Y/X (bytes 0,1); tile and attrs survive. */
void ZeroObjectPositions(void)
{
	wOAMOffset = 0;
	for (uint8_t i = 0; i < OAM_COUNT; i++) {
		uint16_t p = (uint16_t)(wOAM_ADDR + (uint16_t)i * 4u);
		gb_write8(p, 0);
		gb_write8((uint16_t)(p + 1u), 0);
	}
}

/* objects.asm:1-42. n==0 means 256 (post-test loop). Exit hl is always
 * &wOAMOffset regardless of path. */
SetManyObjResult SetManyObjectsAttributes(uint16_t hl, uint8_t d, uint8_t e)
{
	uint8_t off = wOAMOffset;
	if (off >= OAM_SIZE)
		return (SetManyObjResult){wOAMOffset_ADDR, 1};

	uint32_t n = gb_read8(hl++);
	n = n ? n : 0x100u;
	uint16_t oam = (uint16_t)(wOAM_ADDR + off);
	uint8_t carry = 0;
	for (uint32_t i = 0; i < n; i++) {
		gb_write8(oam++, (uint8_t)(gb_read8(hl++) + e));
		gb_write8(oam++, (uint8_t)(gb_read8(hl++) + d));
		gb_write8(oam++, gb_read8(hl++));
		gb_write8(oam++, gb_read8(hl++));
		if ((uint8_t)(oam - wOAM_ADDR) >= OAM_SIZE) {
			carry = 1;
			break;
		}
	}
	wOAMOffset = (uint8_t)(oam - wOAM_ADDR);
	return (SetManyObjResult){wOAMOffset_ADDR, carry};
}
